SpatialGDK/Source/SpatialGDK/Private/Interop/Startup/AssignedPartitionStartupSteps.cpp

```diff
@@ -0,0 +1,144 @@
+{
+	TArray<TUniquePtr<FStartupStep>> Steps;
+
+	Steps.Emplace(MakeUnique<FReadVirtualWorkerTranslatorStep>(State, LocalPartitionId, NetDriver->Connection->GetCoordinator(),
+															   *NetDriver->GlobalStateManager));
+	Steps.Emplace(MakeUnique<FWaitForAssignedPartition>(State, LocalPartitionId, *NetDriver, NetDriver->Connection->GetCoordinator()));
+
+	return Steps;
+}
+
+FReadVirtualWorkerTranslatorStep::FReadVirtualWorkerTranslatorStep(TSharedRef<FInternalState> InState,
+																   TSharedRef<TOptional<Worker_PartitionId>> InLocalPartitionIdPtr,
+																   ISpatialOSWorker& InConnection,
+																   UGlobalStateManager& InGlobalStateManager)
+	: State(InState)
+	, LocalPartitionIdPtr(InLocalPartitionIdPtr)
+	, Connection(&InConnection)
+	, GlobalStateManager(&InGlobalStateManager)
+{
+	StepName = TEXT("Retrieving VirtualWorkerTranslator assigned partition");
+}
+
+bool FReadVirtualWorkerTranslatorStep::TryFinish()
+{
+	const ComponentData* DeploymentMapComponent =
+		Connection->GetView()[SpatialConstants::INITIAL_GLOBAL_STATE_MANAGER_ENTITY_ID].Components.FindByPredicate(
+			ComponentIdEquality{ SpatialConstants::DEPLOYMENT_MAP_COMPONENT_ID });
+	if (ensure(DeploymentMapComponent != nullptr))
+	{
+		const int32 ExistingSessionId = Schema_GetInt32(DeploymentMapComponent->GetFields(), SpatialConstants::DEPLOYMENT_MAP_SESSION_ID);
+
+		GlobalStateManager->ApplySessionId(ExistingSessionId);
+	}
+
+	const EntityViewElement* VirtualWorkerTranslatorEntity =
+		Connection->GetView().Find(SpatialConstants::INITIAL_VIRTUAL_WORKER_TRANSLATOR_ENTITY_ID);
+	if (VirtualWorkerTranslatorEntity != nullptr)
+	{
+		const ComponentData* WorkerTranslatorComponentData = VirtualWorkerTranslatorEntity->Components.FindByPredicate(
+			ComponentIdEquality{ SpatialConstants::VIRTUAL_WORKER_TRANSLATION_COMPONENT_ID });
+		if (WorkerTranslatorComponentData != nullptr)
+		{
+			// We've received worker translation data.
+			TMap<VirtualWorkerId, SpatialVirtualWorkerTranslator::WorkerInformation> VirtualWorkerMapping;
+			SpatialVirtualWorkerTranslator::ApplyMappingFromSchema(VirtualWorkerMapping, *WorkerTranslatorComponentData->GetFields());
+			for (const auto& Mapping : VirtualWorkerMapping)
+			{
+				if (Mapping.Value.ServerWorkerEntityId == State->WorkerEntityId)
+				{
+					// We've received a virtual worker mapping that mentions our worker entity ID.
+					*LocalPartitionIdPtr = Mapping.Value.PartitionEntityId;
+					return true;
+				}
+			}
+		}
+	}
+
+	return false;
+}
+
+FWaitForAssignedPartition::FWaitForAssignedPartition(TSharedRef<FInternalState> InState,
+													 TSharedRef<const TOptional<Worker_PartitionId>> InLocalPartitionIdPtr,
+													 USpatialNetDriver& InNetDriver, ISpatialOSWorker& InConnection)
+	: State(InState)
+	, LocalPartitionIdPtr(InLocalPartitionIdPtr)
+	, NetDriver(&InNetDriver)
+	, Connection(&InConnection)
+{
+	StepName = TEXT("Waiting for assigned partition to come into view");
```

Comment:
Random thought - could make this either the constructor for FStartupStep, or we don't actually make this a member field and just have a virtual function for GetName.

Reply:
The constructor one might ensure the new steps do this but realistically I'm not sure this is something that actually matters.